advancedtls: Add SNI logic to ServerOptions.GetCertificate

cindyxue opened the pull request:
This PR does the following thing:

For go1.10 and above, `config.GetCertificate` is now able to perform SNI logic based on `ServerOptions.GetCertificate` and return the certificate that matches the common name or DNS field in the given `ClientHelloInfo`.

For go1.9, `config.GetCertificate` will always return the first certificate in the certificate list given by `ServerOptions.GetCertificate` because `SupportsCertificate` is not supported.

@ZhenLian

ZhenLian commented on security/advancedtls/advancedtls.go:
```diff
@@ -25,6 +25,7 @@ import (
 	"context"
 	"crypto/tls"
 	"crypto/x509"
+
```
delete this empty line.

ZhenLian commented on security/advancedtls/advancedtls.go:
```diff
@@ -164,7 +165,7 @@ type ServerOptions struct {
 	// client when a new connection is established. This is known as peer
 	// certificate reloading.
-	GetCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+	GetCertificate func(tls.ClientHelloInfo) ([]tls.Certificate, error)
```
Let's keep the pointers to avoid unintentional copying by value.

cindyxue replied:
Sure!

ZhenLian commented on security/advancedtls/advancedtls.go:
```diff
@@ -475,3 +479,27 @@ func cloneTLSConfig(cfg *tls.Config) *tls.Config {
+// GetCertificateWithSNI returns the certificate that matches the SNI field
+// based on the given ClientHelloInfo if GetCertificate returns a list of
+// certificates
+func (o *ServerOptions) GetCertificateWithSNI(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
```
This will become a public API. Can we make it internal?

cindyxue replied:
Does it mean I need to change `GetCertificateWithSNI` to `getCertificateWithSNI`?

ZhenLian replied:
yes.

cindyxue commented:
@ZhenLian Thanks for the review! I made some changes. Could you please check it again?

ZhenLian commented on security/advancedtls/advancedtls.go:
```diff
@@ -164,7 +164,7 @@ type ServerOptions struct {
 	// client when a new connection is established. This is known as peer
 	// certificate reloading.
-	GetCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+	GetCertificate func(*tls.ClientHelloInfo) (*[]tls.Certificate, error)
```
Let's make this a slice of a pointer of certificates, instead of a pointer to a slice.

cindyxue replied:
Sure!

ZhenLian commented on security/advancedtls/advancedtls.go:
```diff
@@ -475,3 +478,28 @@ func cloneTLSConfig(cfg *tls.Config) *tls.Config {
+// GetCertificateWithSNI is a helper function that returns the certificate
+// that matches the SNI field based on the given ClientHelloInfo
+// if GetCertificate returns a list of certificates
+func (o *ServerOptions) getCertificateWithSNI(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
```
Let's make the function signature to be 
```
buildGetCertificateFunc(clientHello *tls.ClientHelloInfo, o *ServerOptions) func(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error)
```
So basically we are trying to make it clear that here we are just trying to build the function instead of actually executing them.

cindyxue replied:
I see, that makes sense!

ZhenLian commented on security/advancedtls/advancedtls.go:
```diff
@@ -475,3 +478,28 @@ func cloneTLSConfig(cfg *tls.Config) *tls.Config {
+	if err != nil {
+		return nil, err
+	}
+	if len(*certificates) == 0 {
```
Is this the right way to check the length of a slice?

cindyxue replied:
Let me double check on this :)

cindyxue replied:
I changed this to `len(certificates)` because now certificates is a slice of ptr to certificates.

menghanl commented:
This PR is still a draft, is it right?

cindyxue commented:
@menghanl Yes :)

cindyxue commented:
@ZhenLian I added a unit test to test the newly added GetCertificate with SNI logic. However, one of the Travis-CI test fails and the error message shows `security/advancedtls/advancedtls.go:526:24: clientHello.SupportsCertificate undefined (type *tls.ClientHelloInfo has no field or method SupportsCertificate)`. According to the golang `tls` package, this function `func (chi *ClientHelloInfo) SupportsCertificate(c *Certificate) error` does exist. Would you know how can I fix this?

ZhenLian commented:
@cindyxue It looks like Go 1.9 doesn't support this function .... To bypass this check, you may want to use the [build Constraints](https://golang.org/pkg/go/build/) to have different implementations on different Go versions. See https://github.com/grpc/grpc-go/pull/3626 for example.

cindyxue commented:
@ZhenLian Yay it works! This PR is ready for review now :)

ZhenLian commented:
As we talked offline, it could be better if we separate out the functions that will have different implementations in different go versions, and add build tag on them, instead of adding build tags on the whole advancedtls files. Can you please do that? Thanks!

cindyxue commented:
@ZhenLian Yes! Now I have separate implementations of `buildGetCertificateFunc` and it's able to pass the travis-ci test. In order to execute files with `// +build sni`, we have to specify `sni` tag when building or testing it.
Eg. `go test ./... -tags sni` and `go build -tags sni`
When I ran `go test ./... -tags sni` locally, both unit test and integration test are able to work. However, since travis-ci tests didn't specify any tag by default, it never went through files with `// +build sni`, which means it didn't run unit test and integration test.

cindyxue commented:
@ZhenLian I updated build constraints for both versions of SNI and it's able to pass all the Travis-CI tests now! Could you please take a look when you have time? Thank you so much :)

ZhenLian commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,105 @@ func TestOptionsConfig(t *testing.T) {
 }
+
+const (
+	tlsEcdheRsaWith3desEdeCbcSha uint16 = 0xc012
```
Why do we have these package-level variables? I think it might be better to have them in the test function?

cindyxue replied:
Ok! I'll move these inside the test function.

ZhenLian commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,105 @@ func TestOptionsConfig(t *testing.T) {
+			if err != nil {
+				t.Fatalf("Unable to generate serverConfig. Error: %v", err)
+			}
+			// "foo.bar.com" is the common name on server certificate server_cert_1.pem.
```
For this, can you please also document in the README file that is used to generate the certificates?

cindyxue replied:
Sure! Would you want me to add an appendix section and list the common name for each certificate there?

ZhenLian replied:
Yes, please! I just thought we probably wanted to document the purposes of all the test certs, since the number could grow larger as time goes. You can do that for the certs you added and I will add them for the rest.

cindyxue replied:
Sounds good!

ZhenLian commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,105 @@ func TestOptionsConfig(t *testing.T) {
+		},
+	}
+	for _, test := range tests {
+		test := test
```
For testing, I would suggest the following: we set three certificates when setting the callback(since now we can set a list of certs), and in the clientHello we specify one of them. Then we use serverConfig.GetCertificate() to check if the one we want is selected. WDYT?

cindyxue replied:
I changed the structure of test cases a bit. Could you please check it again and see if that's what you meant? Thank you!

ZhenLian commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,86 @@ func TestOptionsConfig(t *testing.T) {
+	)
+	const (
+		VersionTLS10 uint16 = 0x0301
+	)
```
Can we remove the const declarations here?

cindyxue replied:
I removed these three const declarations and instead called the corresponding exported const variable from tls library directly.
```	
        const (
		tlsEcdheRsaWith3desEdeCbcSha uint16 = 0xc012
	)
	const (
		CurveP256 tls.CurveID = 23
	)
	const (
		VersionTLS10 uint16 = 0x0301
	)
```
However, for `pointFormatUncompressed`, it's an unexported const in tls library so I have to re-declare it. A reference of the code can be found [here](https://github.com/golang/go/blob/cd10f5f632983e63b156478abdfa0bd57e93898c/src/crypto/tls/common.go#L145-L147).

ZhenLian commented on security/advancedtls/testdata/server_cert_3.pem:
```diff
@@ -0,0 +1,22 @@
+-----BEGIN CERTIFICATE-----
```
Several comments for the generated certificate:
1. we might want to set longer validity for the test certificate, maybe 20 years or so. Otherwise it could cause some unexpected problems when it's expired after a year.
2. Ideally we want to see SNI works for both Common Name and the DNS field under SAN. Since we already had certificates with some common names, can we set some DNS names under the SAN field of this certificate? To learn the difference of these two, you can refer to: [this link](https://support.dnsimple.com/articles/what-is-common-name/#:~:text=on%20Google%20Safari-,Common%20Name%20vs%20Subject%20Alternative%20Name,wildcard%20or%20non%2Dwildcard%20name.&text=The%20SAN%20allows%20issuance%20of,Authority%20and%20the%20specific%20product.)

cindyxue replied:
Sure, I can generate another certificate with DNS names.

cindyxue replied:
Update: removed the old certificate pair `server_cert_3.pem` and `server_key_3.pem` which was a certificate with common name setting only. Generated a new certificate pair with the same name `server_cert_3.pem` and `server_key_3.pem`, but had both common name and DNS names set.
In addition, I also added a new test case to test if SNI works for DNS field of the newly added certificate pair.

easwars replied:
Might be worth adding the human readable version of the cert (as generated by the openssl command) either to the same file (if it supports comments), or add it as a separate file. Something like `server_cert_3.txt`.

cindyxue commented on security/advancedtls/testdata/README.md:
````diff
@@ -40,3 +40,18 @@ commands we run:
    $ openssl verify -verbose -CAfile ca_cert.pem  subject_cert.pem
 
    ```
+
````
No worries! Let me delete these.

cindyxue replied:
I can add one more line that explains how to parse certificate in terminal if you think that's helpful!
FYI, this is the command I used: `openssl x509 -noout -text -in server_cert_3.pem`

cindyxue commented:
@ZhenLian Thanks for the code review! Fixed issues addressed in the comments. Could you please review it again and see if there are more places to change? Thank you so much :D

ZhenLian commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,85 @@ func TestOptionsConfig(t *testing.T) {
 }
+
+func TestGetCertificateSNI(t *testing.T) {
+	const (
```
I think it's quite uncommon to have a const inside a function ... directly assign a 0 value doesn't work? You can also make it a local variable and enforce its type.

cindyxue replied:
Sure, I'll change it to a local variable. Just thought it's more clear and consistent to assign a variable name that matched with other files.

ZhenLian commented:
@easwars Hi Easwar, Can you please take a look at this PR please? Thanks in advance!

easwars commented on security/advancedtls/advancedtls.go:
```diff
@@ -166,7 +166,7 @@ type ServerOptions struct {
 	// client when a new connection is established. This is known as peer
 	// certificate reloading.
-	GetCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+	GetCertificate func(*tls.ClientHelloInfo) ([]*tls.Certificate, error)
```
Should this be `GetCertificates` instead, now that multiple certs can be returned?

cindyxue replied:
Yes! Good catch :)

easwars commented on security/advancedtls/sni_110.go:
```diff
@@ -0,0 +1,53 @@
+
+// The function buildGetCertificateFunc returns the certificate that matches the SNI field
+// for the given ClientHelloInfo, defaulting to the first element of o.GetCertificate.
+func buildGetCertificateFunc(clientHello *tls.ClientHelloInfo, o *ServerOptions) (*tls.Certificate, error) {
```
This function is named `buildGetCertificateFunc`, but is not returning a `func`. Could you please either change the implementation to return a func, or change the name.

cindyxue replied:
Sure, I'll change the name to `buildGetCertificates`

easwars commented on security/advancedtls/advancedtls.go:
```diff
@@ -234,9 +234,16 @@ func (o *ServerOptions) config() (*tls.Config, error) {
+		Certificates: o.Certificates,
+	}
+	// The function getCertificateWithSNI is only able to perform SNI logic for go1.10 and above.
+	// It will return the first certificate in o.GetCertificate for go1.9.
```
This comment would be better served if placed at the `GetCertificate` function comment. Users should know what to expect by reading the godoc.

cindyxue replied:
Sure!

easwars commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,83 @@ func TestOptionsConfig(t *testing.T) {
+		testdata.Path("server_key_3.pem"))
+	if err != nil {
+		t.Fatalf("Server is unable to parse peer certificates. Error: %v", err)
+	}
```
Maybe add a newline to separate the loading of the certs from the test case definitions.

cindyxue replied:
Sounds good

easwars commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,83 @@ func TestOptionsConfig(t *testing.T) {
+
+func TestGetCertificateSNI(t *testing.T) {
+	// Load server certificates for setting the serverGetCert callback function.
+	serverPeerCert1, err := tls.LoadX509KeyPair(testdata.Path("server_cert_1.pem"),
```
`serverPeerCert1` is really the server's certificate, right? Then renaming it as `serverCert1` sounds more appropriate.

cindyxue replied:
There are two types of certificates, peer certificate and trust certificate. We are loading server's peer certificates here, which is why I named it to `serverPeerCert1`. I could change the comment to `"Load server peer certificates for setting the serverGetCert callback function."` if that seems more clear to you!

easwars replied:
I understand that there are two types of certificates. One establishes your identity (called identityCert or just cert), and the other helps validate the certs provided by the other end (called trustCerts, trustRoots or simply roots). We usually call the other end as the peer. So, for the server, the client is the peer and vice-versa. HTH.

cindyxue replied:
I see, that makes sense. Thanks for the clarification!

easwars commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,83 @@ func TestOptionsConfig(t *testing.T) {
+		t.Fatalf("Server is unable to parse peer certificates. Error: %v", err)
+	}
+	serverPeerCert2, err := tls.LoadX509KeyPair(testdata.Path("server_cert_2.pem"),
+		testdata.Path("server_key_2.pem"))
```
Please do not split lines like this. Go does not impose line length restrictions, and one generally splits up lines only when it helps with readability.

cindyxue replied:
Got it, thanks for letting me know :)

easwars commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,83 @@ func TestOptionsConfig(t *testing.T) {
+	serverPeerCert1, err := tls.LoadX509KeyPair(testdata.Path("server_cert_1.pem"),
+		testdata.Path("server_key_1.pem"))
+	if err != nil {
+		t.Fatalf("Server is unable to parse peer certificates. Error: %v", err)
```
Prefer identifying the function and the inputs in the error message. So, this would look something like:

```
t.Fatalf("tls.LoadX509KeyPair(server_cert_1.pem, server_key_1.pem) failed: %v", err)
```

cindyxue replied:
Good point!

easwars commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,83 @@ func TestOptionsConfig(t *testing.T) {
+		desc                string
+		serverGetCert       func(*tls.ClientHelloInfo) ([]*tls.Certificate, error)
+		serverName          string
+		expectedCertificate tls.Certificate
```
s/expectedCertificate/wantCert.

`want` is preferred over `expected` in identifier names in tests since it is shorter and conveys the same meaning.

cindyxue replied:
Ok!

easwars commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,83 @@ func TestOptionsConfig(t *testing.T) {
+		expectedCertificate tls.Certificate
+	}{
+		{
+			desc: "Selected certificate by SNI should be serverPeerCert1",
```
The `desc` field which is passed as the first argument to `t.Run` is used to generate the Go subtest name. So, this subtest would get the following name: `TestGetCertificateSNI/Selected_certificate_by_SNI_should_be_serverPeerCert1`. So, instead of using complete sentences here, think more of these in terms of identifier names and use short and descriptive names. Thanks.

cindyxue replied:
Will do!

easwars commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,83 @@ func TestOptionsConfig(t *testing.T) {
+			}
+			pointFormatUncompressed := uint8(0)
+			clientHello := &tls.ClientHelloInfo{
+				CipherSuites:      []uint16{tls.TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA},
```
Are the other fields outside of `ServerName` useful for this test? If not, you can get rid of them.

cindyxue replied:
Yes, they are required to return the wanted result from `SupportsCertificate`. More details of the function logic can be found [here](https://github.com/golang/go/blob/d286e61b6787fe2b55bf0ec8a814962ebda8d202/src/crypto/tls/common.go#L1006-L1154) :)

easwars commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,83 @@ func TestOptionsConfig(t *testing.T) {
+			}
+			serverConfig, err := serverOptions.config()
+			if err != nil {
+				t.Fatalf("Unable to generate serverConfig. Error: %v", err)
```
Same comment here. Identify the function and the inputs in error messages.

cindyxue replied:
Sure!

easwars commented on security/advancedtls/advancedtls_test.go:
```diff
@@ -679,3 +681,83 @@ func TestOptionsConfig(t *testing.T) {
+			if err != nil {
+				t.Fatalf("Server is unable to parse peer certificates. Error: %v", err)
+			}
+			if !reflect.DeepEqual(*gotCertificate, test.expectedCertificate) {
```
Please prefer using `cmp.Equal` instead of `reflect.DeepEqual`. The former is a safer option.

cindyxue replied:
This comment didn't show up in "files changed" tab for some reason, so I'll reply here.
I tried to use `cmp.Equal`, but I'm getting this error `cannot handle unexported field at {tls.Certificate}.PrivateKey.(*rsa.PrivateKey).PublicKey.N.neg`. Would you know what caused it by any chance? Or we couldn't use `cmp.Equal` for comparing certificates?

easwars replied:
You can use `cmpopts.IgnoreUnexported` to fix this issue.

easwars commented on security/advancedtls/advancedtls_integration_test.go:
```diff
@@ -1,3 +1,5 @@
+// +build go1.10
```
Does this mean that the go1.9 functionality is verified in this integration test? Is it possible to not add this build constraint?

easwars replied:
Github can't seem to keep track of the association between comments and the corresponding file/line-number when you do a force commit. Unfortunate.

easwars replied:
But don't you want to verify the functionality for go1.9 as well?

cindyxue replied:
I'll add testing for go1.9 separately. Thanks for bringing it up!

cindyxue commented:
@easwars Thank you so much for the code review, Easwar! I made changes addressing the review comments. Could you please take another look and see if there are more places to change? Thanks again!

easwars commented on security/advancedtls/advancedtls.go:
```diff
@@ -234,9 +234,16 @@ func (o *ServerOptions) config() (*tls.Config, error) {
+		Certificates: o.Certificates,
+	}
+	if o.GetCertificates != nil {
+		getCertificateWithSNI := func(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
```
Nit: You can inline this function definition along with the assignment to `config.GetCertificate`.

easwars commented on security/advancedtls/sni_110.go:
```diff
@@ -0,0 +1,53 @@
+	"fmt"
+)
+
+// The function buildGetCertificates returns the certificate that matches the SNI field
```
Nit: Skip the leading `The function` in the comment. Go comments style for functions is to begin with the function name.

easwars commented on security/advancedtls/sni_before_110.go:
```diff
@@ -0,0 +1,41 @@
+	"fmt"
+)
+
+// The function buildGetCertificates returns the first element of o.GetCertificates.
```
Same here.

easwars commented on security/advancedtls/sni_test_110.go:
```diff
@@ -0,0 +1,108 @@
+			if err != nil {
+				t.Fatalf("serverConfig.GetCertificate(clientHello) failed: %v", err)
+			}
+			if !cmp.Equal(gotCertificate, test.wantCert, cmp.AllowUnexported(tls.Certificate{}, tls.Certificate{})) {
```
Why is `tls.Certificate{}` mentioned twice?

cindyxue replied:
Removed the extra one.

easwars commented on security/advancedtls/sni_test_before_110.go:
```diff
@@ -0,0 +1,108 @@
+			if err != nil {
+				t.Fatalf("serverConfig.GetCertificate(clientHello) failed: %v", err)
+			}
+			if !cmp.Equal(gotCertificate, test.wantCert, cmp.AllowUnexported(tls.Certificate{}, tls.Certificate{})) {
```
Same here, about repeating `tls.Certificate{}`.

cindyxue replied:
Removed.

ZhenLian commented:
Thanks @cindyxue for the implementation and Thanks @easwars for the review! Merging now ...
